src/general/world_effects/PhotosynthesisProductionEffect.cs

```diff
@@ -114,27 +114,57 @@ private void ApplyCompoundsAddition()
                         {
-                            oxygenBalance += output.Value * outputModifier * effectiveSpeed * species.Value;
+                            oxygenProduced += output.Value * outputModifier * effectiveSpeed * species.Value;
                         }
                         else if (output.Key.ID is Compound.Carbondioxide)
                         {
-                            co2Balance += output.Value * outputModifier * effectiveSpeed * species.Value;
+                            co2Produced += output.Value * outputModifier * effectiveSpeed * species.Value;
                         }
                     }
                 }
             }
 
-            // Scale the balances to make the changes less drastic
-            oxygenBalance *= modifier;
-            co2Balance *= modifier;
+            patch.Biome.TryGetCompound(Compound.Oxygen, CompoundAmountType.Biome, out var existingOxygen);
+            patch.Biome.TryGetCompound(Compound.Carbondioxide, CompoundAmountType.Biome, out var existingCo2);
+
+            // Unless something else changes it, compound values stay the same
+            float oxygenTarget = existingOxygen.Ambient;
+            float co2Target = existingCo2.Ambient;
+
+            float total = existingOxygen.Ambient + existingCo2.Ambient;
+
+            // Special (but common) case where zero oxygen is being produced
+            if (oxygenProduced == 0 && oxygenConsumed > 0)
+            {
+                oxygenTarget = 0.0f;
+
+                if (co2Produced > 0)
+                {
+                    co2Target = total;
+                }
+            }
+
+            // Special (but common) case where zero carbon dioxide is being produced
+            if (co2Produced == 0 && co2Consumed > 0)
+            {
+                co2Target = 0.0f;
 
-            changesToApply.Clear();
+                if (oxygenProduced > 0)
+                {
+                    oxygenTarget = total;
+                }
+            }
 
-            if (oxygenBalance != 0)
-                changesToApply[Compound.Oxygen] = oxygenBalance;
+            // if both compounds are being produced, calculate an aproximate steady state value
+            if (oxygenProduced > 0 && co2Produced > 0)
+            {
+                // Calculate long-term equilibrium balances based on production and consumption ratio
+                oxygenTarget = oxygenProduced / (oxygenConsumed + co2Consumed + MathUtils.EPSILON) * total;
+                co2Target = co2Produced / (oxygenConsumed + co2Consumed + MathUtils.EPSILON) * total;
+            }
 
-            if (co2Balance != 0)
-                changesToApply[Compound.Carbondioxide] = co2Balance;
+            changesToApply[Compound.Oxygen] = (oxygenTarget - existingOxygen.Ambient) * 0.5f;
+            changesToApply[Compound.Carbondioxide] = (co2Target - existingCo2.Ambient) * 0.5f;
 
             if (changesToApply.Count > 0)
```

Comment:
This if no longer works correctly after your changes because even if the change to apply is less than EPSILON the value is added to the dictionary so this check no longer works. Either it needs to be reworked that small values (so close to target) are not processed or this needs to loop here to check if any of the changes are above EPSILON. I think the first approach will be easier to do and clearer to understand here.